### PR TITLE
Change the way the TickerTask tracks Blocks

### PR DESCRIPTION
**Pull request by md5sha256**
## Description

As per a discussion in the SF discord, this PR removes the use of explicit locks used in TickerTask. 
I'm not a massive fan of the implementation of `TickingBlocks` as it is right now. Currently its just been wrapped (mutex) so it isn't that much better for performance than it was before (might even be worse on that end). However, we now map the world to a long as opposed to `ChunkPosition --> Set<Location>` which means the memory footprint is considerably lower. 

## Proposed changes

- Add `TickingBlocks`, a class that serves as an accessor/wrapper for a `Map<World, Set<Long>>`. In other words, a `BlockPositionSet`. 
- Removes the public getters for the `tickingLocations`
- Swap out the `Lock` + `Map<ChunkPosition, Set<Location>>` for `TickingBlocks`
- Added relevant docs

## Checklist

- [ ] I have fully tested the proposed changes and promise that they will not break everything into chaos.
- [ ] I have also tested the proposed changes in combination with various popular addons and can confirm my changes do not break them.
- [x] I followed the existing code standards and didn't mess up the formatting.
- [x] I did my best to add documentation to any public classes or methods I added.
- [x] I have added `Nonnull` and `Nullable` annotations to my methods to indicate their behaviour for null values
- [ ] I added sufficient Unit Tests to cover my code.

### CODE REVIEW

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickingBlocks.java [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+import java.util.Set;
+import java.util.WeakHashMap;
+
+/**
+ * Memory efficient, synchronized collection of {@link BlockPosition BlockPositions} used by the
+ * {@link TickerTask}. All methods are thread-safe unless explicitly stated otherwise.
+ *
+ * @author md5sha256
+ */
+final class TickingBlocks {
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickingBlocks.java:21]: Not a fan of this name tbh.
Perhaps `TickerCollection`?

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickingBlocks.java [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+     * Get the compressed {@link BlockPosition BlockPositions} mapped to a given {@link World}.
+     * The returned {@link Set} is thread-safe. Changes to set will be reflected in this instance.
+     *
+     * @param world The world
+     * @return Returns a {@link Set} of {@link Long Longs} which are compressed {@link BlockPosition BlockPositions}
+     *
+     * @see #getBlocksCopy()
+     */
+    public @Nonnull Set<Long> getBlocks(@Nonnull World world) {
+        return blocks.computeIfAbsent(world, (unused) -> Collections.synchronizedSet(new HashSet<>()));
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickingBlocks.java:35]: Don't put braces around a single variable in a lambda expression
```suggestion
        return blocks.computeIfAbsent(world, x -> Collections.synchronizedSet(new HashSet<>()));
```

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickingBlocks.java [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+        }
+        positions.remove(block.getPosition());
+        if (positions.isEmpty()) {
+            blocks.remove(world);
+        }
+    }
+
+    public void removeBlock(@Nonnull Location location) {
+        removeBlock(new BlockPosition(location));
+    }
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickingBlocks.java:125]: sum javadocs? :D

(No necessity, just nice to have)

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickingBlocks.java [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+     */
+    public @Nonnull Set<BlockPosition> getBlocksCopy() {
+        Set<BlockPosition> positions = new HashSet<>();
+        for (Map.Entry<World, Set<Long>> entry : blocks.entrySet()) {
+            World world = entry.getKey();
+            Set<BlockPosition> converted = convertToBlockPosition(world, entry.getValue());
+            positions.addAll(converted);
+        }
+        return positions;
+    }
```
Comment (TheBusyBiscuit) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickingBlocks.java:69]: These methods seem to be unused and are public, yet this class isn't even public and the TickerTask provides no getter for this object anyway.

**Review by TheBusyBiscuit — CHANGES_REQUESTED** [review]

**Comment by md5sha256** [comment]
I've resolved the comments. That being said I've also changed the implementation of `TickingCollection` to use a guava `Cache` as that offers better concurrent performance whilst also supporting weak keys. Alongside biscuit, could @WalshyDev also have a look through?

**Comment by TheBusyBiscuit** [comment]
> Alongside biscuit, could @WalshyDev also have a look through?

I am not enough :FeelsBadMan:

**Comment by md5sha256** [comment]
> > Alongside biscuit, could @WalshyDev also have a look through?
> 
> I am not enough :FeelsBadMan:

Sorry :P Just thought Walshy's experience with the guava Cache might make a difference since we've discussed how to implement this for a bit back in April and early May

**Comment by md5sha256** [comment]
👀

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerCollection.java [hunk 5]
```diff
@@ -0,0 +1,129 @@
...
+import java.util.Set;
+import java.util.WeakHashMap;
+
+/**
+ * Memory efficient, synchronized collection of {@link BlockPosition BlockPositions} used by the
+ * {@link TickerTask}. All methods are thread-safe unless explicitly stated otherwise.
+ *
+ * @author md5sha256
+ */
+final class TickerCollection {
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerCollection.java:23]: We should annotate with @ThreadSafe

**Review by WalshyDev — COMMENTED** [review]
Looking good. Has this been stress tested?

(also apologies, didn't realise this was waiting on me)

**Comment by md5sha256** [comment]
No, it has not. I can start working on a JMH bench. I just needed an opinion on the use of the Guava Cache as I feel like I'm abusing it here.

**Comment by WalshyDev** [comment]
Hmmm,Guava's Cache is made and optimised for automatic eviction (usually size or time). We don't really use any of the Cache's features at all nor use it for the intended purpose. I feel like we're better off just using ConcurrentHashMap.

The only feature/benefit we get from Cache is the weakKeys. Considering the compute and retrieval of things isn't expensive and memory is more valuable to MC servers than usual production software I think we stick to native Java ConcurrentHashMap.

**Comment by md5sha256** [comment]
Noted. So should I wrap all keys with a weak reference?

**Comment by WalshyDev** [comment]
Probably best yeah

**Comment by md5sha256** [comment]
I guess I have to stick with holding the UUID instance of the world instead?

**Comment by WalshyDev** [comment]
Yeah UUID would be better.

**Comment by md5sha256** [comment]
Also, by stress test do you mean a unit test that stresses the collection to check for concurrency errors or a benchmark?

**Comment by WalshyDev** [comment]
> Also, by stress test do you mean a unit test that stresses the collection to check for concurrency errors or a benchmark?

Making sure it can handle concurrency correctly.

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerCollection.java [hunk 6]
```diff
@@ -0,0 +1,134 @@
+package io.github.thebusybiscuit.slimefun4.implementation.tasks;
+
+
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerCollection.java:3]: remove double new line

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerTask.java [hunk 7]
```diff
@@ -75,10 +66,11 @@
      */
     private static final long MAX_WAIT_TIME = 1_500_000L;
 
+
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerTask.java:69]: double new line

**Comment by md5sha256** [comment]
Should be good EXCEPT for https://discord.com/channels/565557184348422174/573872826327629844/856183240736899104. 
The tests currently pass as I've modified the test to not fail (I changed it to assert against a fresh `Set` from `#getBlocks` instead of `compressedPositions`. The stress test passes; however, I'm all eyes and ears for whether the stress test needs more work. 
Tested on an Intel i7-7700 (4 cores, 8 threads).

**Comment by md5sha256** [comment]
Also, I was forced to merge in `master` because I couldn't compile the new spelling tests as the localisation stuff changed. Oh well.

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerCollection.java [hunk 8]
```diff
@@ -0,0 +1,133 @@
...
+            return;
+        }
+        positions.remove(block.getPosition());
+        if (positions.isEmpty()) {
+            blocks.remove(world.getUID());
+        }
+    }
+
+    /**
+     * Remove a location from this ticker collection.
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerCollection.java:98]:
```suggestion
     * Remove a {@link Location} from this ticker collection.
```

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerCollection.java [hunk 9]
```diff
@@ -0,0 +1,133 @@
...
+     *
+     * @param location The location of the block to remove
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerCollection.java:100]:
```suggestion
     * @param location The {@link Location} of the block to remove
```

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerCollection.java [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+     * @see #removeBlock(BlockPosition)
+     */
+    public void removeBlock(@Nonnull Location location) {
+        removeBlock(new BlockPosition(location));
+    }
+
+    /**
+     * Clear all the blocks for a given {@link World}
+     * @param key The world to clear the key for
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerCollection.java:109]:
```suggestion
     * Clear all the blocks for a given {@link World}
     *
     * @param key The {@link World} to remove blocks from
```

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerTask.java [hunk 11]
```diff
@@ -101,17 +92,19 @@
...
     private final ExecutorService asyncExecutor = Executors.newSingleThreadExecutor(threadFactory);
 
+
     /**
-     * Making collections concurrent doesn't make their iterators thread safe.
+     *  These are "Queues" of blocks that need to be removed or moved.
+     *  <p>
+     *  Whilst the {@link ConcurrentHashMap} implementation is thread-safe,
+     *  the {@link Location Locations} are not. Therefore, the internal copy of the locations
+     *  should be treated as <strong>read only</strong> and never exposed externally.
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerTask.java:106]: 2 spaces at the start here
```suggestion
     * These are "Queues" of blocks that need to be removed or moved.
     * <p>
     * Whilst the {@link ConcurrentHashMap} implementation is thread-safe,
     * the {@link Location Locations} are not. Therefore, the internal copy of the locations
     * should be treated as <strong>read only</strong> and never exposed externally.
```

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerTask.java [hunk 12]
```diff
@@ -338,22 +320,21 @@ private void processAsyncTasks() {
...
-                if (item != null && item.getBlockTicker() != null) {
-                    try {
-                        runTask(l, data, item);
-                        tickers.add(item.getBlockTicker());
-                    } catch (Exception x) {
-                        // Catch and report any errors caused by this ticker.
-                        reportErrors(l, item, x);
-                    }
+    /**
+     * Executes all tasks in a given {@link Collection} of locations
```
Comment (WalshyDev) [inline src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerTask.java:324]:
```suggestion
     * Executes all tasks in a given {@link Collection} of {@link Location locations}
```

**Review by WalshyDev — CHANGES_REQUESTED** [review]

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerCollection.java [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+     * Get a copy of all the blocks stored in the form of {@link BlockPosition BlockPositions}
+     * The returned {@link Set} is not guaranteed to be thread-safe.
+     *
+     * @return Returns a {@link Set} of {@link BlockPosition BlockPositions}
+     */
+    public @Nonnull Set<Location> getBlocksAsLocations() {
+        Set<Location> locations = new HashSet<>();
+        blocks.keySet().removeIf(uid -> Bukkit.getWorld(uid) == null);
+        for (Map.Entry<UUID, Set<Long>> entry : blocks.entrySet()) {
+            World world = Bukkit.getWorld(entry.getKey());
```
[the remark this answers is not in the document]
Comment (md5sha256) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerCollection.java:51]: Won't that throw a concurrent modification exception, when removing keys whilst iterating over the entryset?

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerCollection.java [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+    }
+
+    private static @Nonnull Set<Location> convertToLocation(@Nonnull World world, @Nonnull Collection<Long> raw) {
+        Set<Location> locations = new HashSet<>(raw.size());
+        for (long compressed : raw) {
+            int x = (int) (compressed >> 38);
+            int y = (int) (compressed & 0XFFF);
+            int z = (int) (compressed << 26 >> 38);
+            locations.add(new Location(world, x, y, z));
+        }
```
[the remark this answers is not in the document]
Comment (md5sha256) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerCollection.java:129]: I'll PR that to cscl2 and wait for a bump.

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerCollection.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (WalshyDev) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerCollection.java:51]: Nope, it's thread-safe.

[the remark this answers is not in the document]
Comment (md5sha256) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerCollection.java:51]: I guess, but won't that also cause the original iterator the for-loop is using to have undertiministic behaviour?

[the remark this answers is not in the document]
Comment (WalshyDev) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerCollection.java:51]: Hmmm yeah fair enough.

File: src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerCollection.java (hunk 14 above)
[the remark this answers is not in the document]
Comment (md5sha256) [reply src/main/java/io/github/thebusybiscuit/slimefun4/implementation/tasks/TickerCollection.java:128]: Righty then, now waiting on https://github.com/TheBusyBiscuit/CS-CoreLib2/pull/194

**Review by WalshyDev — APPROVED** [review]
LGTM

**Review by TheBusyBiscuit — APPROVED** [review]
What Walshy said 🤷🏻

**Pull request merged**